zerver/lib/test_helpers.py

```diff
@@ -598,3 +601,98 @@ def zulip_reaction_info() -> Dict[str, str]:
+        for pattern in valid_line_patterns:
+            self.compiled_line_patterns.append(re.compile(pattern))
+        self.full_spam = ""
+
+    def find_spam(self, data: str) -> None:
+        lines = data.split('\n')
+        for line in lines:
+            if not line:
+                continue
+            found_spam = True
+            for compiled_pattern in self.compiled_line_patterns:
+                if compiled_pattern.match(line):
+                    found_spam = False
+                    break
+            if found_spam:
+                self.full_spam += f'{line}\n'
+
+class TeeStderrAndFindSpam():
+    def __init__(self, spam_finder: SpamFinder) -> None:
+        self.stderr_stream = sys.stderr
+
+        # get shared console handler instance from any logger that have it
+        self.console_log_handler = cast(logging.StreamHandler, logging.getLogger('django.server').handlers[0])
+
+        assert isinstance(self.console_log_handler, logging.StreamHandler)
+        assert self.console_log_handler.stream == sys.stderr
+        self.spam_finder = spam_finder
+
+    def __enter__(self) -> None:
+        sys.stderr = self  # type: ignore[assignment] # Doing tee by swapping stderr stream with custom file like class
+        self.console_log_handler.stream = self  # type: ignore[assignment] # Doing tee by swapping stderr stream with custom file like class
+
+    def __exit__(self, exc_type: Optional[Type[BaseException]],
+                 exc_value: Optional[BaseException],
+                 traceback: Optional[TracebackType]) -> None:
+        sys.stderr = self.stderr_stream
+        self.console_log_handler.stream = sys.stderr
+
+    def write(self, data: str) -> None:
+        self.stderr_stream.write(data)
+        self.spam_finder.find_spam(data)
+
+    def writelines(self, data: Iterable[str]) -> None:
+        self.stderr_stream.writelines(data)
+        lines = "".join(data)
+        self.spam_finder.find_spam(lines)
+
+    def flush(self) -> None:
+        self.stderr_stream.flush()
+
+class TeeStdoutAndFindSpam():
+    def __init__(self, spam_finder: SpamFinder) -> None:
+        self.stdout_stream = sys.stdout
+        self.spam_finder = spam_finder
+
+    def __enter__(self) -> None:
+        sys.stdout = self  # type: ignore[assignment] # Doing tee by swapping stderr stream with custom file like class
+
+    def __exit__(self, exc_type: Optional[Type[BaseException]],
+                 exc_value: Optional[BaseException],
+                 traceback: Optional[TracebackType]) -> None:
+        sys.stdout = self.stdout_stream
+
+    def write(self, data: str) -> None:
+        self.stdout_stream.write(data)
+        self.spam_finder.find_spam(data)
+
+    def writelines(self, data: Iterable[str]) -> None:
+        self.stdout_stream.writelines(data)
+        lines = "".join(data)
+        self.spam_finder.find_spam(lines)
+
+    def flush(self) -> None:
+        self.stdout_stream.flush()
```

Comment:
You might be able to deduplicate these two classes; they're nearly the same.  Not sure if it's worth it.